buildSrc/subprojects/parser/src/main/java/org/apache/calcite/buildtools/parser/DialectGenerate.java

```diff
@@ -0,0 +1,416 @@
+   * @param tokens The tokens starting from the function declaration and
+   *               ending at EOF
+   * @param charIndex The character index of the entire text of the file at which
+   *                  the parsing is commencing at
+   */
+  private int processCurlyBlock(
+      StringBuilder functionBuilder,
+      Queue<String> tokens,
+      int charIndex) {
+    // Remove any preceeding spaces or new lines before the curly block starts.
+    charIndex = consumeExtraSpacesAndLines(functionBuilder, tokens, charIndex);
+    if (!tokens.peek().equals("{")) {
+      System.out.println(tokens);
+      throw new IllegalStateException("First token of curly block must be a curly brace.");
+    }
+    CurlyParser parser = new CurlyParser();
+    while (!tokens.isEmpty()) {
+      String token = tokens.poll();
+      functionBuilder.append(token);
+      charIndex += token.length();
+      if (parser.parseToken(token)) {
+        return charIndex;
+      }
+    }
+    return charIndex;
+  }
+
+  /**
+   * Consumes tokens while they are either a space or line break. This needs
+   * to be called before a curly block begins being parsed to remove any extra
+   * spaces or new lines that may preceed it.
+   *
+   * @param functionBuilder The builder unto which the tokens get added to once parsed
+   * @param tokens The tokens starting from charIndex until EOF
+   * @param charIndex The character index of the entire text of the file at which
+   *                  the parsing is commencing at
+   */
+  private int consumeExtraSpacesAndLines(
+      StringBuilder functionBuilder,
+      Queue<String> tokens,
+      int charIndex) {
+    while (tokens.peek().equals(" ") || tokens.peek().equals("\n")) {
+      String token = tokens.poll();
+      charIndex += token.length();
+      functionBuilder.append(token);
+    }
+    return charIndex;
+  }
+
+  /**
+   * Gets the function name from the declaration.
+   *
+   * @param functionDeclaration The function declaration of form
+   *                            <return_type> <name> (<args>) :
+   */
+  private String getFunctionName(String functionDeclaration) {
+    Matcher m = namePattern.matcher(functionDeclaration);
+    // Name is the second match.
+    m.find();
+    m.find();
+    return m.group();
+  }
+
+  /**
+   * Responsible for parsing a block of text surrounded by curly braces. It is
+   * assumed that parsing begins at the start of a curly block. Maintains the
+   * state of which structure the parser is currently in to ensure that
+   * encountered curly braces are valid.
+   */
+  static class CurlyParser {
+
+    /**
+     * All of the possible important structures that a token can be inside of.
+     * This is tracked to ensure that when a curly brace is encountered,
+     * whether it is part of the syntax or not can be determined.
+     */
+    enum InsideState {
+
+      /**
+       * Not inside of anything.
+       */
+      NONE,
```

Comment:
This is kind of a misleading name as it can be either not inside anything or inside curly braces but not inside the other states.